File: csharp/1641.count-sorted-vowel-strings.cs
/*
 * @lc app=leetcode id=1641 lang=csharp
 *
 * [1641] Count Sorted Vowel Strings
 */

// @lc code=start
public class Solution {

    private static int CountVowelStrings(int n, int i, int c, int?[,] memo) {
        if (i == n) {
            return 1;
        }
        if (memo[i, c].HasValue) {
            return memo[i, c].Value;
        }
        int ret = 0;
        for (int nc = c; nc < 5; ++nc) {
            ret += CountVowelStrings(n, i + 1, nc, memo);
        }
        memo[i, c] = ret;
        return ret;
    }

    public int CountVowelStrings(int n) {
        return CountVowelStrings(n, 0, 0, new int?[n, 5]);
    }
}
// @lc code=end

File: java/1641.count-sorted-vowel-strings.java
/*
 * @lc app=leetcode id=1641 lang=java
 *
 * [1641] Count Sorted Vowel Strings
 */

// @lc code=start
class Solution {
    public int countVowelStrings(int n) {
        return countVowelStrings(0, n, new Integer[5][n]);
    }

    private static int countVowelStrings(int c, int l, Integer[][] memo) {
        if (l == 0) {
            return 1;
        }
        Integer m = memo[c][l - 1];
        if (m != null) {
            return m;
        }
        int ret = 0;
        for (int nc = c; nc < 5; ++nc) {
            ret += countVowelStrings(nc, l - 1, memo);
        }
        memo[c][l - 1] = ret;
        return ret;
    }
}
// @lc code=end

